examples/trimerge-sync-basic-client/src/WebsocketRemote.ts

```diff
@@ -1,83 +1,124 @@
+    this.socket = new WebSocket(this.websocketUrl);
     this.socket.onopen = () => {
       if (!this.socket) {
-        console.warn(`[TRIMERGE-SYNC] Connected, but already shutdown...`);
+        this.logger?.warn(`[TRIMERGE-SYNC] Connected, but already shutdown...`);
         return;
       }
-      console.log(`[TRIMERGE-SYNC] Connected to ${websocketUrl}`);
-      onEvent({ type: 'remote-state', connect: 'online' });
-      const events = this.bufferedEvents;
-      this.bufferedEvents = [];
+      this.logger?.log(`[TRIMERGE-SYNC] Connected to ${this.websocketUrl}`);
+      this.emit({ type: 'remote-state', connect: 'online' });
+      const events = this.bufferedEventsToServer;
+      this.bufferedEventsToServer = [];
       for (const event of events) {
         this.send(event);
       }
     };
     this.socket.onclose = () => this.fail('closed', 'disconnected');
     this.socket.onerror = () => this.fail('error', 'network', false);
     this.socket.onmessage = (event) => {
-      onEvent(JSON.parse(event.data));
+      this.emit(JSON.parse(event.data));
     };
+
+    const { localStoreId } = await this.commitRepo.dbInfo;
+
     this.send({
       type: 'init',
       version: 1,
-      localStoreId,
       lastSyncCursor,
-      auth,
+      localStoreId,
+      auth: this.auth,
     });
   }
 
-  send(event: SyncEvent<CommitMetadata, Delta, Presence>): void {
-    if (!this.socket) {
-      return;
-    }
-    if (this.socket.readyState === WebSocket.CONNECTING) {
-      this.bufferedEvents.push(event);
-    } else {
-      this.socket.send(JSON.stringify(event));
+  listen(cb: OnRemoteEventFn<CommitMetadata, Delta, Presence>): void {
+    invariant(!this.onEvent, 'listen() called twice');
+    this.onEvent = cb;
+
+    if (this.bufferedEventsToClient.length > 0) {
+      const events = this.bufferedEventsToClient;
+      this.bufferedEventsToClient = [];
+      for (const event of events) {
+        this.emit(event);
+      }
     }
+    this.bufferedEventsToClient = [];
   }
 
-  shutdown(): void {
-    if (!this.socket) {
+  get active() {
+    return this.socket?.readyState === WebSocket.OPEN;
+  }
+
+  disconnect(): void | Promise<void> {
+    if (
+      !this.socket ||
+      this.socket.readyState === WebSocket.CLOSING ||
+      this.socket.readyState === WebSocket.CLOSED
+    ) {
       return;
     }
-    if (this.socket.readyState !== WebSocket.CLOSED) {
-      console.log(
-        `[TRIMERGE-SYNC] Shutting down websocket ${this.socket.url}...`,
-      );
-      this.socket.close(1000, 'shutdown');
-    }
+    this.logger?.log(`Shutting down websocket ${this.socket.url}...`);
+    this.socket.close(1000, 'shutdown');
     this.socket = undefined;
-    this.onEvent({
+    this.emit({
       type: 'remote-state',
       connect: 'offline',
       read: 'offline',
     });
   }
 
+  private emit(event: SyncEvent<CommitMetadata, Delta, Presence>) {
+    if (this.onEvent) {
```

Comment:
Leaving onEvent dangling around for someone to call feels like a bug waiting to happen.  If emit is the only consumer of onEvent, what about emit being a lambda we overwrite? The initial value would be to shove events in the buffer, calling listen would flush the buffer and replace emit with onEvent.

Reply:
FWIW, this is just the example code so nothing in our app actually uses this.